text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+    unsafe fn grow(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Allocates a new handle to a block of zeroed memory. On success, invalidates any existing handle.
+    fn allocate_zeroed(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+
+    /// Grows the block of memory associated to a handle with zeroed memory. On success, the handle is invalidated.
+    unsafe fn grow_zeroed(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+}
+```
+
+_Note:  full-featured documentation of the trait and methods can be found in the companion repository at_
+        https://github.com/matthieu-m/store/blob/main/src/interface.rs.
+
+The `Store` trait is supplemented by 3 additional marker traits, providing extra guarantees:
+
+```rust
+/// A refinement of `Store` which does not invalidate existing handles on allocation, and does not invalidate
+/// unrelated existing handles on deallocation.
+pub unsafe trait MultipleStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers on allocation, resolution, or deallocation, but
+/// may invalidate them on moves.
+pun unsafe trait StableStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers, not even on moves. That is, this refinement
+/// guarantees that the blocks of memory are pinned in memory.
+pub unsafe trait PinningStore: StableStore {}
+```
+
+
+##  Library Organization
+
+This RFC proposes to follow the lead of the `Allocator` trait, and add the `Store` traits to the `core` crate, either in
+the `alloc` module or in a new `store` module.
+
+It leaves to a follow-up RFC the introduction of a `store` or `store-collections` crate which would contain the code of
+the various standard collections: `Box`, `Vec`, `String`, `BinaryHeap`, `BTreeMap`, `BTreeSet`, `LinkedList`, and
+`VecDeque`, all adapted for the `Store` API.
+
+Those types would then be re-exported as-is in the `alloc` crate, drastically reducing its size.
+
+
+#   Drawbacks
+
+This RFC increases the surface of the standard library, with yet another `Allocator`.
+
+Furthermore, the natural consequence of adopting this RFC would be rewriting the existing collections in terms of
+`Store`, rather than `Allocator`. A mostly mechanical task, certainly, but an opportunity to introduce subtle bugs in
+the process, even if MIRI would hopefully catch most such bugs.
```

Comment:
```suggestion
the process, even if Miri would hopefully catch most such bugs.
```
`Miri` is the official spelling (see the title on https://github.com/rust-lang/miri)